WordPress/gutenberg
# Migrate block editor insert usage to preferences store

Part of https://github.com/WordPress/gutenberg/issues/31965

## What?
The block editor package maintains some block insertion data to aid the 'Most used blocks' feature in the inserter. It persists this data.

This PR migrates the data to the preferences store.

## Why?
As outlined in https://github.com/WordPress/gutenberg/issues/31965, the goal is to move persisted data out of our various stores to a centralized preferences store.

## How?
<s>This required adding the preferences store as a dependency of the block editor package.</s>

The logic in the block editor's `preferences` reducer is now moved over to a new action `updateInsertUsage` that uses the preferences store. This new private action is dispatched in the `insertBlocks` and `replaceBlocks` actions.

The local storage data is migrated across to live under the preferences store from the old preferences system.

The preferences store's API usually uses a `scope` parameter that's usually set to 'core/edit-post' or 'core/edit-site', to indicate which editor the data belongs to. `insertUsage` stats are shared across editors, so I've used the 'core' scope.

Finally, this PR introduces a new private action in the preferences store, `markNextChangeAsExpensive`, which can be dispatched before updating a preference to delay the regularity with which the persistence layer makes HTTP requests to update user meta. It means you can insert loads of blocks quickly, but user meta will only update after a minute has passed. Insert usage still backs up immediately to a local storage persistence layer, so there's lots of resilience.

## Testing Instructions
1. Using `trunk` open the post editor
2. In the editor preferences modal, in the blocks section, enable the 'Show most used blocks' setting
3. Delete your WP local storage data and reload
4. Open the main inserter and click one of the blocks repeatedly (not paragraph, which is already first) in the Most Used section until it becomes the first block (note - possibly a bug that the order of blocks update while the inserter is open)
5. Checkout this branch and rebuild
6. Reload the post editor and open the main inserter, the order of the most used blocks should be retained (local storage data was migrated correctly)
7. Select a different block several times and make it the most used block.
8. Reload the post editor and open the main inserter, the order of the most used blocks should be retained (data is still persisted)
